### PR TITLE
Checkout at the specified location

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
The snsdemo needs to be checked out outside the nns-dapp repository.

# Changes
- In the snsdemo checkout action, accept a `path` variable, similar to the vanilla checkout action.
- Improve logging in the snsdemo checkout action.
- Fix the error handling and improve logging in the error cases.

# Tests
- See CI for current use
- For checking the case where snsdemo is already checked out at the requested commit, see: https://github.com/dfinity/nns-dapp/actions/runs/6821083632/job/18550979415
- For the case where snsdemo is already checked out but at the wrong commit: https://github.com/dfinity/nns-dapp/actions/runs/6829080386/job/18574535473

# Todos

- [x] Add entry to changelog (if necessary).

### CODE REVIEW

File: .github/actions/checkout_snsdemo/action.yaml [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
-    out at the expected commit.  If not, this fails with an error merssage.
+    out at the expected commit.  If not, this fails with an error message.
   * Otherwise: Clones the snsdemo repository at the commit specified in `dfx.json`
 inputs:
   token:
     description: "Github access token used to clone"
     required: true
+  path:
+    description: "Where to check out the repository"
+    default: "snsdemo"
```
Comment (bitdivine) [inline .github/actions/checkout_snsdemo/action.yaml:14]: Alternative considered: Set the default to `../snsdemo`.  Decided against it as it has surprise factor for anyone used to the standard checkout action.  Better make the path explicit where the action is called, I think.

File: .github/actions/checkout_snsdemo/action.yaml [hunk 2]
```diff
@@ -35,9 +38,10 @@ runs:
       shell: bash
       run: |
         expected_commit="$(git rev-parse "${{ steps.snsdemo_ref.outputs.ref }}" || true)"
-        actual_commit="$(git rev-parse HEAD)"
+        actual_commit="$(cd "${{ inputs.path }}" && git rev-parse HEAD)"
```
Comment (bitdivine) [inline .github/actions/checkout_snsdemo/action.yaml:41]: The code on the left actually omits the cd which is a bug.

**Review by lmuntaner — COMMENTED** [review]
How are you going to use this? What's the default commit? I can't find it...

**Comment by bitdivine** [comment]
> How are you going to use this?

The action is used in `.github/actions/build_nns_dapp/action.yaml` and `.github/workflows/deploy-to-app.yaml`.  The places where the action is used is not expected to change.

The bulk of this PR is actually bug fixes and they stand on their own even if nothing else is changed.  At root, the bugs were caused by not using the same path everywhere, so making it a variable showed me the bugs.  The original motivation was however to move the snsdemo to somewhere outside the nns-dapp repo and I do still intend to do this, to address some of the many warnings we are getting,  but in a separate PR.

> What's the default commit?

The default  commit is unchanged.  It's described in the action description at the top of the file:
```
Otherwise: Clones the snsdemo repository at the commit specified in `dfx.json`
```

**Review by lmuntaner — APPROVED** [review]
Thanks for the explanations.

Is this related to the comment you made that the snsdemo is used for two different things: utilities and the snapshot?

**Comment by bitdivine** [comment]
> Is this related to the comment you made that the snsdemo is used for two different things: utilities and the snapshot?

The checkouts are used to get the utilities.  This PR changes nothing in that regard.

**Pull request merged**